lib/upipe-framers/upipe_s337_framer.c

```diff
@@ -0,0 +1,437 @@
+ * @param uprobe structure used to raise events
+ * @param signature signature of the pipe allocator
+ * @param args optional arguments
+ * @return pointer to upipe or S337F in case of allocation error
+ */
+static struct upipe *upipe_s337f_alloc(struct upipe_mgr *mgr,
+                                      struct uprobe *uprobe,
+                                      uint32_t signature, va_list args)
+{
+    struct upipe *upipe = upipe_s337f_alloc_void(mgr, uprobe, signature, args);
+    if (unlikely(upipe == NULL))
+        return NULL;
+
+    struct upipe_s337f *upipe_s337f = upipe_s337f_from_upipe(upipe);
+    upipe_s337f_init_urefcount(upipe);
+    upipe_s337f_init_output(upipe);
+    upipe_s337f_init_flow_def(upipe);
+    upipe_s337f->uref = NULL;
+    upipe_throw_ready(&upipe_s337f->upipe);
+    return &upipe_s337f->upipe;
+}
+
+/** @internal @This finds the position of the s337m sync word in the frame
+ */
+static ssize_t upipe_s337f_sync(struct upipe *upipe, struct uref *uref)
+{
+    struct upipe_s337f *upipe_s337f = upipe_s337f_from_upipe(upipe);
+
+    size_t size = 0;
+    uref_sound_size(uref, &size, NULL);
+
+    const int32_t *in;
+    uref_sound_read_int32_t(uref, 0, -1, &in, 1);
+
+    for (size_t i = 0; i < size; i++)
+        if ((in[2*i+0] == (0x6f872 << 12) && in[2*i+1] == (0x54e1f << 12)) ||
+                ((in[2*i+0] == (0x96f872 << 8) && in[2*i+1] == (0xa54e1f << 8)))) {
+            uref_sound_unmap(uref, 0, -1, 1);
+            return i;
+        }
+
+    uref_sound_unmap(uref, 0, -1, 1);
+
+    return -1;
+}
+
+/** @internal
+ *
+ * @param upipe description structure of the pipe
+ * @param uref uref structure
+ * @param upump_p reference to pump that generated the buffer
+ */
+static void upipe_s337f_input(struct upipe *upipe, struct uref *uref, struct upump **upump_p)
+{
+    struct upipe_s337f *upipe_s337f = upipe_s337f_from_upipe(upipe);
+    struct uref *output = upipe_s337f->uref;
+
+    const ssize_t sync = upipe_s337f_sync(upipe, uref);
+
+    if (sync == -1) {
+        if (output) {
+            upipe_err(upipe, "Sync lost");
+            uref_free(output);
+            upipe_s337f->uref = NULL;
+        }
+
+        struct uref *flow_def = upipe_s337f_alloc_flow_def_attr(upipe);
+        if (unlikely(flow_def == NULL)) {
+            upipe_throw_fatal(upipe, UBASE_ERR_ALLOC);
+        } else {
+            const char *def = NULL;
+            UBASE_FATAL(upipe, uref_flow_get_def(upipe_s337f->flow_def_input, &def));
+            UBASE_FATAL(upipe, uref_flow_set_def_va(flow_def, def));
+            flow_def = upipe_s337f_store_flow_def_attr(upipe, flow_def);
+            if (flow_def)
+                upipe_s337f_store_flow_def(upipe, flow_def);
+        }
+        upipe_s337f_output(upipe, uref, upump_p);
+        return;
+    } else if (!output) {
+        upipe_notice(upipe, "Found synchro");
+    }
+
+    if (output) {
+        size_t size[2];
+        uref_sound_size(uref, &size[0], NULL);
```

Comment:
This can fail and segfault miserably later.

Reply:
I guess I should also remove the abort() based error handling

Reply:
yes